### PR TITLE
[SR-11184] Delay binding of key path application projected result value 

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
Delay binding the projected result value binding of the key path application constraint until we've tried the key path type.

Resolves SR-11184.

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [inline lib/Sema/CSSimplify.cpp:5104]: I'm not sure this is the right fix – it seems like the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

I think we'll want to add some logic in `getPotentialBindings` for a `KeyPathApplication` constraint similar to the logic we have for `ApplicableFunction` – where if we notice that the type variable is part of the callsite (on the left hand side), we mark the binding as being `FullyBound`, meaning that we'll try to delay binding the type variable until we've bound other type variables. This allows us to e.g try resolving the overload before attempting to bind the result type of a function.

In this case though, it looks like we only need to check whether the type variable appears in the `getThirdType()` of the `KeyPathApplication` (the projected result type), and if so mark the binding as `FullyBound`. This will allow us to wait until we've tried the key path type before attempting to bind the result type.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:5104]: Thanks for the review @hamishknight !

> I'm not sure this is the right fix – it seems like the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

Yeah, I was really unsure about this approach ... but couldn't really think of anything else 

> the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

That makes sense, and that is probably why we get a fixed type early on this case ... what doesn't happen for non-IUO optionals


> I think we'll want to add some logic in getPotentialBindings for a KeyPathApplication constraint similar to the logic we have for ApplicableFunction – where if we notice that the type variable is part of the callsite (on the left hand side), we mark the binding as being FullyBound, meaning that we'll try to delay binding the type variable until we've bound other type variables. This allows us to e.g try resolving the overload before attempting to bind the result type of a function.
In this case though, it looks like we only need to check whether the type variable appears in the getThirdType() of the KeyPathApplication (the projected result type), and if so mark the binding as FullyBound. This will allow us to wait until we've tried the key path type before attempting to bind the result type.

Great, I'll take a look on that, thanks again for the detailed comment :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Fixed @hamishknight, seems like I needed to better understand how the bindings work in those situations. 
Thanks again for the explanation :)

File: lib/Sema/CSBindings.cpp
Comment (hamishknight) [inline lib/Sema/CSBindings.cpp:555]: Could you add a bit to this comment saying something like:

```
mark the result as FullyBound to ensure we delay binding until
we've bound other type variables. In particular, this ensures we
try to bind the key path type first, which can allow us to 
discover additional bindings for the result type. 
```

Just to clarify the use of `FullyBound`, which IMO has a pretty confusing name.

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: Thanks @LucianoPAlmeida! I have one more comment but this LGTM!

File: lib/Sema/CSBindings.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSBindings.cpp:555]: Sure, I agree... TBH I didn't know the concept, had to look on the comments around the code to understand what this means and how the bindings work.

Comment (hamishknight) [reply lib/Sema/CSBindings.cpp:555]: Yeah the current binding logic is pretty confusing, we're hoping to refactor it to be simpler and more efficient :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSBindings.cpp]: Just added! :)

Comment (hamishknight) [reply lib/Sema/CSBindings.cpp]: Awesome, thank you!

**Review by hamishknight — APPROVED** [review]
I'll let Pavel have the final say, but this looks good to me!

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by LucianoPAlmeida** [comment]
> I'll let Pavel have the final say, but this looks good to me!

Great, thank you @hamishknight!

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by LucianoPAlmeida** [comment]
Thank you @xedin :)

**Comment by xedin** [comment]
FYI, in general if there is a bind constraint that fails, it means that we'd need to adjust something in the way bindings are inferred :)

**Comment by LucianoPAlmeida** [comment]
> FYI, in general if there is a bind constraint that fails, it means that we'd need to adjust something in the way bindings are inferred :)

Right, that makes sense! Thank you @xedin :)

**Pull request merged**
